common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -135,8 +135,70 @@ class LibraryContentModule(LibraryContentFields, XModule, StudioEditableModule):
     any particular student.
     """
 
+    def __init__(self, *args, **kwargs):
+        self.children = []
+        self._selected_set = None
+        super(LibraryContentModule, self).__init__(*args, **kwargs)
+
+    @classmethod
+    def make_selection(cls, selected, children, max_count, mode):
+        """
+        Dynamically selects block_ids indicating which of the possible children are displayed to the current user.
+
+        Arguments:
+            selected - list of (block_type, block_id) tuples assigned to this student
+            children - children of this block
+            max_count - number of components to display to each student
+            mode - how content is drawn from the library
+
+        Returns:
+            A dict containing the following keys:
+
+            'selected' (set) of (block_type, block_id) tuples assigned to this student
+            'invalid' (set) of dropped (block_type, block_id) tuples that are no longer valid
+            'overlimit' (set) of dropped (block_type, block_id) tuples that were previously selected
+            'added' (set) of newly added (block_type, block_id) tuples
+        """
+        selected = set(tuple(k) for k in selected)  # set of (block_type, block_id) tuples assigned to this student
+
+        # Determine which of our children we will show:
+        valid_block_keys = set([(c.block_type, c.block_id) for c in children])
+        # Remove any selected blocks that are no longer valid:
+        invalid_block_keys = (selected - valid_block_keys)
+        if invalid_block_keys:
+            selected -= invalid_block_keys
+
+        # If max_count has been decreased, we may have to drop some previously selected blocks:
+        overlimit_block_keys = set()
+        while len(selected) > max_count:
+            overlimit_block_keys.add(selected.pop())
+
+        # Do we have enough blocks now?
+        num_to_add = max_count - len(selected)
+
+        added_block_keys = None
+        if num_to_add > 0:
+            # We need to select [more] blocks to display to this user:
+            pool = valid_block_keys - selected
+            if mode == "random":
+                num_to_add = min(len(pool), num_to_add)
+                added_block_keys = set(random.sample(pool, num_to_add))
+                # We now have the correct n random children to show for this user.
+            else:
+                raise NotImplementedError("Unsupported mode.")
+            selected |= added_block_keys
+
+        return {
+            'selected': selected,
+            'invalid': invalid_block_keys,
+            'overlimit': overlimit_block_keys,
+            'added': added_block_keys,
+        }
```

Comment:
The edX style guide would suggest using a namedtuple for dictionaries with a static format like this.  Something like
    
    _SelectedBlock = namedtuple('SelectedBlock', ['selected', 'invalid', 'overlimit', 'added'])  # at module level
    return _SelectedBlock(selected, invalid_block_keys, overlimit_block_keys, added_block_keys)